src/Layer.js

```diff
@@ -147,29 +124,51 @@ export default class Layer {
      * @api
      */
     addTo (map, beforeLayerID) {
-        if (this._isCartoMap(map)) {
-            this._addToCartoMap(map, beforeLayerID);
-        } else if (this._isMGLMap(map)) {
-            this._addToMGLMap(map, beforeLayerID);
-        } else {
-            throw new CartoValidationError('layer', 'nonValidMap');
+        const STYLE_ERROR_REGEX = /Style is not done loading/;
+
+        try {
+            map.addLayer(this, beforeLayerID);
+        } catch (error) {
+            if (!STYLE_ERROR_REGEX.test(error)) {
+                throw new Error(error);
+            }
+
+            map.on('load', () => {
+                map.addLayer(this, beforeLayerID);
+            });
         }
     }
 
+    /**
+     * Remove this layer from a map.
+     *
+     * @param {mapboxgl.Map} map - The map on which to remove the layer
+     * @memberof carto.Layer
+     * @instance
+     * @api
+     */
+    removeFrom (map) {
+        map.removeLayer(this);
+    }
+
     /**
      * Update the layer with a new Source and a new Viz object, replacing the current ones. The update is done atomically, i.e.: the viz will be changed with the source, not before it.
      * This method will return a promise that will be resolved once the source and the visualization are validated.
      * The promise will be rejected if the validation fails, for example because the visualization expects a property name that is not present in the source.
      * The promise will be rejected also if this method is invoked again before the first promise is resolved.
      * If the promise is rejected the layer's source and viz won't be changed.
-     * @param {carto.source.Base} source - the new Source object
-     * @param {carto.Viz} viz - the new Viz object
+     * @param {carto.source.Base} source - The new Source object
+     * @param {carto.Viz?} viz - Optional. The new Viz object
      * @memberof carto.Layer
-     * @async
      * @instance
+     * @async
      * @api
      */
     async update (source, viz) {
+        if (viz === undefined) {
+            // Use current viz
+            viz = this._viz;
```

Comment:
I don't remember well but I think this wasn't done before for a reason, this should be tested well

Reply:
I found an example using `layer.getViz()` as a second argument, so this could be the default behaviour. I'll add tests for this.